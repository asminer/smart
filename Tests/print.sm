
// $Id$ 
	
/*

Tests for print
	
*/
	
int one := 1;

real ten2n(int n) := cond (n==0, 1.0, cond (n<0, 1.0 / ten2n(-n), 10.0*ten2n(n-1)));

print("This should be like pascal, ", one, " would hope\n");

print("Here's a test of the width specifier for strings\n");

for (int i in {1..10}) {
  print("XXX":i, "\n");
}

print("Here's a test of the width specifier for integers\n");

for (int i in {1..10}) {
  print(one:i, "\n");
}

print("Testing tabs\n");

print("N\tN^2\tN^3\n");
print("--\t---\t----\n");

for (int i in {1..11}) {
  print(i, "\t", i*i, "\t", i*i*i, "\n");
}

print("Using SCIENTIFIC\n");

# OutputRealFormat SCIENTIFIC

print("Width and precision specifiers for reals\n");
for (int i in {-9..9}) {
  print(5+ten2n(i):20:10+i, "\n");
}

print("Only width specifier for reals\n");
for (int i in {-9..9}) {
  print(5+ten2n(i):20+i, "\n");
}

print("Using FIXED\n");

# OutputRealFormat FIXED

print("Width and precision specifiers for reals\n");
for (int i in {-9..9}) {
  print(5+ten2n(i):20:10+i, "\n");
}

print("Only width specifier for reals\n");
for (int i in {-9..9}) {
  print(5+ten2n(i):20+i, "\n");
}

print("Using GENERAL\n");

# OutputRealFormat GENERAL
print("Precision specifiers for reals\n");
for (int i in {1..9}) {
  print(i, ": ", 5+ten2n(6):1:i, "\n");
}

print("Precision specifiers for reals without second ':field'\n");
for (int i in {1..9}) {
  print(i, ": ", 5+ten2n(6):1, "\n");
}

print("Examples for the manual\n");
# OutputRealFormat FIXED
print(123.8:4:2,"\n");
print(123.8:12:2,"\n");
print(123.87654321:12,"\n");
print(123.87654321,"\n");
# OutputRealFormat SCIENTIFIC
print(123.8:4:2,"\n");
print(123.8:12:2,"\n");
print(123.87654321:12,"\n");
print(123.87654321,"\n");
# OutputRealFormat GENERAL
print(123.8:4:2,"\n");
print(123.8:12:8,"\n");
print(123.87654321:12,"\n");
print(123.87654321,"\n");

real x := 123.8;
# OutputRealFormat FIXED
print("FIXED\n");
for (int p in {2..7}) {
  print("p=",p,"\t");
  for (int w in {4..8}) {
    print("w=",w," [",x:w:p,"]\t");
  }
  print("\n");
}
# OutputRealFormat SCIENTIFIC
print("SCIENTIFIC\n");
for (int p in {2..7}) {
  print("p=",p,"\t");
  for (int w in {4..8}) {
    print("w=",w," [",x:w:p,"]\t");
  }
  print("\n");
}
# OutputRealFormat GENERAL
print("GENERAL\n");
for (int p in {2..7}) {
  print("p=",p,"\t");
  for (int w in {4..8}) {
    print("w=",w," [",x:w:p,"]\t");
  }
  print("\n");
}

print("testing spacing: this is an error","X":5:3,"\n");

print("testing spacing: this","abcdef":5,"should take 6 positions, not 5\n");
print("testing spacing: this","X":5,"should take 5 positions\n");
print("testing spacing: this",2:5,"should take 5 positions\n");
print("testing spacing: this",2.0:5,"should take 5 positions\n");
print("testing spacing: this",2.0:5:3,"should take 5 positions\n");
print("testing spacing: this",2.3333333:5:3,"should take 5 positions\n");

print("New and improved left justification with negative widths:\n");


print("integers\n");
int digits(int n) := cond(n==1, 1, digits(n-1)*10+n);

for(int i in {1..9}) {
  print("left  |", digits(i):-10, "|\n");
}
for(int i in {1..9}) {
  print("right |", digits(i):10, "|\n");
}

print("strings\n");

string letter(int n) := case(n, " ", 1:"a", 2:"b", 3:"c", 4:"d", 5:"e", 6:"f", 7:"g", 8:"h", 9:"i", 10:"j");

string letters(int n) := cond(n==1, "a", letters(n-1)+letter(n));

for(int i in {1..9}) {
  print("left  |", letters(i):-10, "|\n");
}
for(int i in {1..9}) {
  print("right |", letters(i):10, "|\n");
}

print("reals\n");

real decimals(int n) := cond(n==1, 0.1, decimals(n-1)+n/ten2n(n));

for(int i in {1..7}) {
  print("left  |", decimals(i):-10:8, "|\n");
}
for(int i in {1..7}) {
  print("right |", decimals(i):10:8, "|\n");
}


print("Thousand separators for integers:\n");

#OutputThousandSeparator ","

for (int i in {9..1..-1}) {
  print(-digits(i), "\n");
}
for (int i in {1..9}) {
  print(digits(i), "\n");
}

print("Thousand separators with width:\n");

for(int i in {7..1..-1}) {
  print("left  |", -digits(i):-10, "|\n");
}
for(int i in {1..8}) {
  print("left  |", digits(i):-10, "|\n");
}

for(int i in {7..1..-1}) {
  print("right |", -digits(i):10, "|\n");
}
for(int i in {1..8}) {
  print("right |", digits(i):10, "|\n");
}

