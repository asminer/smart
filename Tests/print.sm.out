This should be like pascal, 1 would hope
Here's a test of the width specifier for strings
XXX
XXX
XXX
 XXX
  XXX
   XXX
    XXX
     XXX
      XXX
       XXX
Here's a test of the width specifier for integers
1
 1
  1
   1
    1
     1
      1
       1
        1
         1
Testing tabs
N	N^2	N^3
--	---	----
1	1	1
2	4	8
3	9	27
4	16	64
5	25	125
6	36	216
7	49	343
8	64	512
9	81	729
10	100	1000
11	121	1331
Using SCIENTIFIC
Width and precision specifiers for reals
             5.0e+00
            5.00e+00
           5.000e+00
          5.0000e+00
         5.00001e+00
        5.000100e+00
       5.0010000e+00
      5.01000000e+00
     5.100000000e+00
    6.0000000000e+00
   1.50000000000e+01
  1.050000000000e+02
 1.0050000000000e+03
1.00050000000000e+04
1.000050000000000e+05
1.0000050000000000e+06
1.00000050000000000e+07
1.000000050000000000e+08
1.0000000050000000000e+09
Only width specifier for reals
5.000000e+00
5.000000e+00
 5.000000e+00
  5.000001e+00
   5.000010e+00
    5.000100e+00
     5.001000e+00
      5.010000e+00
       5.100000e+00
        6.000000e+00
         1.500000e+01
          1.050000e+02
           1.005000e+03
            1.000500e+04
             1.000050e+05
              1.000005e+06
               1.000000e+07
                1.000000e+08
                 1.000000e+09
Using FIXED
Width and precision specifiers for reals
                 5.0
                5.00
               5.000
              5.0000
             5.00001
            5.000100
           5.0010000
          5.01000000
         5.100000000
        6.0000000000
      15.00000000000
    105.000000000000
  1005.0000000000000
10005.00000000000000
100005.000000000000000
1000005.0000000000000000
10000005.00000000000000000
100000005.000000000000000000
1000000005.0000000000000000000
Only width specifier for reals
   5.000000
    5.000000
     5.000000
      5.000001
       5.000010
        5.000100
         5.001000
          5.010000
           5.100000
            6.000000
            15.000000
            105.000000
            1005.000000
            10005.000000
            100005.000000
            1000005.000000
            10000005.000000
            100000005.000000
            1000000005.000000
Using GENERAL
Precision specifiers for reals
1: 1e+06
2: 1e+06
3: 1e+06
4: 1e+06
5: 1e+06
6: 1e+06
7: 1000005
8: 1000005
9: 1000005
Precision specifiers for reals without second ':field'
1: 1e+06
2: 1e+06
3: 1e+06
4: 1e+06
5: 1e+06
6: 1e+06
7: 1e+06
8: 1e+06
9: 1e+06
Examples for the manual
123.80
      123.80
  123.876543
123.876543
1.24e+02
    1.24e+02
1.238765e+02
1.238765e+02
1.2e+02
       123.8
     123.877
123.877
FIXED
p=2	w=4 [123.80]	w=5 [123.80]	w=6 [123.80]	w=7 [ 123.80]	w=8 [  123.80]	
p=3	w=4 [123.800]	w=5 [123.800]	w=6 [123.800]	w=7 [123.800]	w=8 [ 123.800]	
p=4	w=4 [123.8000]	w=5 [123.8000]	w=6 [123.8000]	w=7 [123.8000]	w=8 [123.8000]	
p=5	w=4 [123.80000]	w=5 [123.80000]	w=6 [123.80000]	w=7 [123.80000]	w=8 [123.80000]	
p=6	w=4 [123.800000]	w=5 [123.800000]	w=6 [123.800000]	w=7 [123.800000]	w=8 [123.800000]	
p=7	w=4 [123.8000000]	w=5 [123.8000000]	w=6 [123.8000000]	w=7 [123.8000000]	w=8 [123.8000000]	
SCIENTIFIC
p=2	w=4 [1.24e+02]	w=5 [1.24e+02]	w=6 [1.24e+02]	w=7 [1.24e+02]	w=8 [1.24e+02]	
p=3	w=4 [1.238e+02]	w=5 [1.238e+02]	w=6 [1.238e+02]	w=7 [1.238e+02]	w=8 [1.238e+02]	
p=4	w=4 [1.2380e+02]	w=5 [1.2380e+02]	w=6 [1.2380e+02]	w=7 [1.2380e+02]	w=8 [1.2380e+02]	
p=5	w=4 [1.23800e+02]	w=5 [1.23800e+02]	w=6 [1.23800e+02]	w=7 [1.23800e+02]	w=8 [1.23800e+02]	
p=6	w=4 [1.238000e+02]	w=5 [1.238000e+02]	w=6 [1.238000e+02]	w=7 [1.238000e+02]	w=8 [1.238000e+02]	
p=7	w=4 [1.2380000e+02]	w=5 [1.2380000e+02]	w=6 [1.2380000e+02]	w=7 [1.2380000e+02]	w=8 [1.2380000e+02]	
GENERAL
p=2	w=4 [1.2e+02]	w=5 [1.2e+02]	w=6 [1.2e+02]	w=7 [1.2e+02]	w=8 [ 1.2e+02]	
p=3	w=4 [ 124]	w=5 [  124]	w=6 [   124]	w=7 [    124]	w=8 [     124]	
p=4	w=4 [123.8]	w=5 [123.8]	w=6 [ 123.8]	w=7 [  123.8]	w=8 [   123.8]	
p=5	w=4 [123.8]	w=5 [123.8]	w=6 [ 123.8]	w=7 [  123.8]	w=8 [   123.8]	
p=6	w=4 [123.8]	w=5 [123.8]	w=6 [ 123.8]	w=7 [  123.8]	w=8 [   123.8]	
p=7	w=4 [123.8]	w=5 [123.8]	w=6 [ 123.8]	w=7 [  123.8]	w=8 [   123.8]	
ERROR in file print.sm near line 124:
	No match for print(string, string:int:int, string)
testing spacing: thisabcdefshould take 6 positions, not 5
testing spacing: this    Xshould take 5 positions
testing spacing: this    2should take 5 positions
testing spacing: this    2should take 5 positions
testing spacing: this    2should take 5 positions
testing spacing: this 2.33should take 5 positions
New and improved left justification with negative widths:
integers
left  |1         |
left  |12        |
left  |123       |
left  |1234      |
left  |12345     |
left  |123456    |
left  |1234567   |
left  |12345678  |
left  |123456789 |
right |         1|
right |        12|
right |       123|
right |      1234|
right |     12345|
right |    123456|
right |   1234567|
right |  12345678|
right | 123456789|
strings
left  |a         |
left  |ab        |
left  |abc       |
left  |abcd      |
left  |abcde     |
left  |abcdef    |
left  |abcdefg   |
left  |abcdefgh  |
left  |abcdefghi |
right |         a|
right |        ab|
right |       abc|
right |      abcd|
right |     abcde|
right |    abcdef|
right |   abcdefg|
right |  abcdefgh|
right | abcdefghi|
reals
left  |0.1       |
left  |0.12      |
left  |0.123     |
left  |0.1234    |
left  |0.12345   |
left  |0.123456  |
left  |0.1234567 |
right |       0.1|
right |      0.12|
right |     0.123|
right |    0.1234|
right |   0.12345|
right |  0.123456|
right | 0.1234567|
Thousand separators for integers:
-123,456,789
-12,345,678
-1,234,567
-123,456
-12,345
-1,234
-123
-12
-1
1
12
123
1,234
12,345
123,456
1,234,567
12,345,678
123,456,789
Thousand separators with width:
left  |-1,234,567|
left  |-123,456  |
left  |-12,345   |
left  |-1,234    |
left  |-123      |
left  |-12       |
left  |-1        |
left  |1         |
left  |12        |
left  |123       |
left  |1,234     |
left  |12,345    |
left  |123,456   |
left  |1,234,567 |
left  |12,345,678|
right |-1,234,567|
right |  -123,456|
right |   -12,345|
right |    -1,234|
right |      -123|
right |       -12|
right |        -1|
right |         1|
right |        12|
right |       123|
right |     1,234|
right |    12,345|
right |   123,456|
right | 1,234,567|
right |12,345,678|
